heartbeat/lvmlockd

```diff
@@ -0,0 +1,438 @@
+		# this daemon up, we expect it to adopt locks from a previous
+		# instance of lvmlockd.
+		extras="$extras -A 1"
+	fi
+	# This client only support "dlm" lock manager
+	extras="$extras -g dlm"
+
+	ocf_log info "starting ${LOCKD}..."
+	ocf_run ${LOCKD} -p ${OCF_RESKEY_pidfile} $extras
+	rc=$?
+	if [ $rc -ne $OCF_SUCCESS ] ; then
+		ocf_exit_reason "Failed to start ${LOCKD}, exit code: $rc"
+		return $OCF_ERR_GENERIC
+	fi
+
+	# lvmlockd requires shared VGs to be started before they're used
+	ocf_log info "start lockspaces of shared VG(s)..."
+	ocf_run vgchange --lockstart $(shared_vgs)
+	rc=$?
+	if [ $rc -ne $OCF_SUCCESS ] ; then
+		ocf_exit_reason "Failed to start shared VG(s), exit code: $rc"
+		return $OCF_ERR_GENERIC
+	fi
+
+	activate_all_vgs
+	return $OCF_SUCCESS
+}
+
+# Each shared VG has its own lockspace. Besides, lvm_global lockspace
+# is for global use, and it should be the last one to close. It should
+# be enough to only check on lvm_global.
+wait_lockspaces_close()
+{
+	local retries=0
+
+	ocf_log info "Waiting for all lockspaces to be closed"
+	while [ $retries -lt "$TIMEOUT_COUNT" ]
+	do
+		if ! dlm_tool ls lvm_global | grep -Eqs "^name[[:space:]]+lvm_global" ; then
+			return $OCF_SUCCESS
+		fi
+
+		sleep 0.5
+		retries=$((retries + 1))
+	done
+
+	ocf_exit_reason "Failed to close all lockspaces clearly"
+	exit $OCF_ERR_GENERIC
+}
+
+kill_stop()
+{
+	local pid=$1
+	local retries=0
+
+	ocf_log info "Killing ${LOCKD} (pid=$pid)"
+	while
+		daemon_is_running $pid && [ $retries -lt "$TIMEOUT_COUNT" ]
+	do
+		if [ $retries -ne 0 ] ; then
+			# don't sleep on the first try
+			sleep 0.5
+		fi
+		kill -s TERM $pid >/dev/null 2>&1
+		retries=$((retries + 1))
+	done
+
+}
+
+lvmlockd_stop() {
+	local pid=""
+
+	if ! silent_status ; then
+		ocf_log info "${LOCKD} is not running"
+		return $OCF_SUCCESS
+	fi
+
+	deactivate_all_vgs
+
+	# lvmlockd requires shared VGs to be started before they're used
+	ocf_log info "stop the lockspaces of shared VG(s)..."
+	ocf_run vgchange --lockstop $(shared_vgs)
+	rc=$?
+	if [ $rc -ne $OCF_SUCCESS ] ; then
+		ocf_exit_reason "Failed to stop VG(s), exit code: $rc"
+		return $OCF_ERR_GENERIC
+	fi
+
+	wait_lockspaces_close
+
+	pid=$(get_pid)
+	kill_stop $pid
+	if silent_status ; then
+		ocf_exit_reason "Failed to stop, ${LOCKD}[$pid] still running."
+		return $OCF_ERR_GENERIC
+	fi
+
+	return $OCF_SUCCESS
+}
+
+lvmlockd_monitor() {
+	if silent_status ; then
+		return $OCF_SUCCESS
+	fi
+
+	ocf_log info "${LOCKD} not running"
+	return $OCF_NOT_RUNNING
+}
+
+lvmlockd_validate() {
+	check_binary ${LOCKD}
+	check_binary vgchange
+	check_binary vgs
+	check_binary lvmconfig
+	check_binary dlm_tool
+	check_binary pgrep
+
+	return $OCF_SUCCESS
+}
+
+
+# Make sure meta-data and usage always succeed
+case $__OCF_ACTION in
+meta-data)		meta_data
+			exit $OCF_SUCCESS
+			;;
+usage|help)		usage
+			exit $OCF_SUCCESS
+			;;
+esac
+
+# Anything other than meta-data and usage must pass validation
+lvmlockd_validate || exit $?
```

Comment:
I'm not sure about this. No other agents run validation for each action, only in the validate-all call. Are you worried that pacemaker does not act correctly?

Reply:
Aha, I learned this pattern from resource agent guide:

[4. Resource agent structure](http://www.linux-ha.org/doc/dev-guides/_execution_block.html)
```

# Anything other than meta-data and usage must pass validation
foobar_validate_all || exit $?
```